packages/actor-context-preprocess-statistic-link-discovery/lib/StatisticLinkDiscovery.ts

```diff
@@ -0,0 +1,84 @@
+// eslint-disable-next-line import/no-nodejs-modules
+import { EventEmitter } from 'events';
+import type { ILink, IDiscoverEventData, IStatisticDiscoveredLinks } from '@comunica/types';
+
+export class StatisticLinkDiscovery implements IStatisticDiscoveredLinks {
+  // Number of discover events tracked
+  public discoverEvents: number;
+
+  // Directed edge list
+  public edgeList: Set<string>;
+
+  // Metadata is saved as follows: First key indicates what url this metadata belongs to, while value is
+  // a list of all metadata objects recorded. This list can contain multiple Records as links can be discovered from
+  // multiple data sources.
+  public metadata: Record<string, Record<any, any>[]>;
+
+  // Emitter that indicates when a new discover statistic is recorded. This can be used downstream for, e.g, aggregate
+  // statistic trackers
+  public statisticEvents: EventEmitter;
+
+  // Logger that emits the recorded data.
+  public log: ((message: string, data?: (() => any)) => void);
+
+  public constructor(log: (message: string, data?: (() => any)) => void) {
+    this.discoverEvents = 0;
+
+    this.edgeList = new Set();
+    this.metadata = {};
+
+    this.statisticEvents = new EventEmitter();
+
+    this.log = log;
+  }
+
+  public setDiscoveredLink(link: ILink, parent: ILink): boolean {
+    // If self-edge or duplicate edge we don't track.
+    if (link.url === parent.url || this.edgeList.has(JSON.stringify([ parent.url, link.url ]))) {
+      return false;
+    }
+
+    this.edgeList.add(JSON.stringify([ parent.url, link.url ]));
+
+    const discoveredLinkMetadata = {
+      ...link.metadata,
+      discoveredTimestamp: Date.now(),
+      discoverOrder: this.discoverEvents,
+    };
+
+    if (discoveredLinkMetadata) {
+      // Retain previous metadata if this link has already been discovered, and add any metadata in the passed link
+      this.metadata[link.url] = this.metadata[link.url] ?
+          [ ...this.metadata[link.url], discoveredLinkMetadata ] :
+          [ discoveredLinkMetadata ];
+    }
+
+    const discoverEventData: IDiscoverEventData = {
+      edge: [ parent.url, link.url ],
+      metadataChild: this.metadata[link.url],
+      metadataParent: this.metadata[parent.url],
+    };
+
+    this.emit(discoverEventData);
+
+    // Log info if available
+    this.log('Discover Event', () => ({
```

Comment:
Is log still needed if we already have emit?
I would suspect logging can be implemented on top of emit.

Reply:
Do you mean by passing a logger to the emit function as an optional parameter?

Reply:
No, not really.
What we can do, is remove this logger code here.
And if the user still requires the logging of this information, the user can add a listener to this statistic, and manually call the logger from the statistic listener callback.
Do you think this make sense?

Reply:
That makes sense, though this would require the user to make an actor. As the statistics are only available in the context, which is inaccessible to the user.

Reply:
> As the statistics are only available in the context, which is inaccessible to the user.

Won't they be available to the user that invoked the query (through the JS API)?
Because the user can provide an input context, and can also obtain the output context after starting execution.

Reply:
They can obtain the output context? Oops never knew, then I completely agree with removing the log function